src/udev/udev-event.c

```diff
@@ -1023,10 +1023,6 @@ int udev_event_execute_rules(UdevEvent *event,
         if (r < 0)
                 return r;
 
-        r = update_devnode(event);
-        if (r < 0)
-                return r;
-
```

Comment:
hmm, I'd really keep this here in place, so that we give the guarantee that the device node is fully set up once we update the db. That's a really important property: the udev db should only be updated once all the other stuff is in place (in particular all the metadata the db carries, and device node info is among that), so that if people enumerate things with the udev db the data stored in there actually matches reality.

Could you explain why you moved this down? given that the symlinks db is separate from the regular device db there should not be a need to move this down?

Reply:
The idea is that `link_find_prioritized()` gets information about link priorities of other devices that claim the same symlink from their database entires. By changing the order we increase the chance that it finds a device with highest priority. Btw, if I don't do this `udev-test.pl` starts failing on test cases that simulate parallel addition of the devices that all want the same symlink (with different priority).

As for guarantees, well, once you receive uevent the db is updated. And even previously, you could get into state where db entry contained data that didn't match the reality since symlinks where updated before db. I think the only guarantee is that for single kernel event, once processed, and client receives device event the DB is consistent. This change doesn't change that guarantee.

Reply:
still, the udev db should reflect not report symlinks that don't exist yet. the udev db should be updated last.

The symlink db is kep separate and is maintained inside of update_devnode(), hence the udev db should not be relevant to create the correct symlink.

not sure i grok the udev-test.pl failure, that's something to track down.

udev is not just about delivering uevents. It's also that when you enumerate stuff (i.e. udev_enumerate, not udev_monitor) things you discover then are already in place.

Thus, the order of the three things should be: first update symlink/devnode, second update the db, third tell everyone about it